File: server/static/sdk-dart/yekonga.io.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

/// Flutter/Dart WebSocket client with Socket.IO-like API
/// Features:
/// - Automatic reconnection with exponential backoff
/// - Message queuing when disconnected
/// - Rooms support
/// - Acknowledgments (acks)
/// - Fluent API: socket.to('room').emit(...)
/// - Events: connect, disconnect, reconnecting, error, etc.

typedef EventCallback = void Function(dynamic data);
typedef AckCallback = void Function(dynamic response);

class YekongaSocket {
  WebSocketChannel? _channel;
  final String url;
  final Map<String, dynamic> _opts;

  final Map<String, List<EventCallback>> _eventListeners = {};
  final Set<String> _rooms = {};
  final List<_PendingMessage> _pendingMessages = [];
  final Map<String, _PendingAck> _pendingAcks = {};

  int _reconnectAttempts = 0;
  Timer? _reconnectTimer;
  bool _connecting = false;

  int _nextMsgId = 0;

  String? id;
  bool connected = false;
  bool disconnected = true;

  YekongaSocket(
    this.url, {
    bool reconnection = true,
    int reconnectionAttempts = double.infinity.toInt(),
    int reconnectionDelay = 1000,
    int reconnectionDelayMax = 5000,
    int timeout = 10000,
  }) : _opts = {
         'reconnection': reconnection,
         'reconnectionAttempts': reconnectionAttempts,
         'reconnectionDelay': reconnectionDelay,
         'reconnectionDelayMax': reconnectionDelayMax,
         'timeout': timeout,
       } {
    connect();
  }

  void connect() {
    if (_connecting) return;
    _connecting = true;

    try {
      final u = Uri.parse(url);
      final socketUrl = Uri.parse(
        '${u.origin}/yekonga.io/?ns=${u.path}',
      ).toString();

      _channel = WebSocketChannel.connect(Uri.parse(socketUrl));

      _channel!.stream.listen(
        (data) {
          try {
            final msg = jsonDecode(data as String);

            // Handle acknowledgment
            if (msg['type'] == 'ack' && msg['msgId'] != null) {
              final pending = _pendingAcks.remove(msg['msgId']);
              if (pending != null) {
                pending.timer.cancel();
                pending.ack(msg['payload'] ?? null);
              }
              return;
            }

            // Client ID from server
            if (msg['event'] == 'id') {
              id = msg['data'];
              _emitLocal('id', id);
              return;
            }

            // Regular event
            if (msg['event'] != null) {
              _emitLocal(msg['event'], msg['data']);
            }
          } catch (err) {
            print('Failed to parse message: $err');
          }
        },
        onDone: () {
          connected = false;
          disconnected = true;
          _connecting = false;
          id = null;

          _emitLocal('disconnect', {
            'code': _channel?.closeCode,
            'reason': _channel?.closeReason,
          });

          if (_opts['reconnection'] == true &&
              _reconnectAttempts < _opts['reconnectionAttempts']) {
            _scheduleReconnect();
          } else if (_reconnectAttempts >= _opts['reconnectionAttempts']) {
            _emitLocal('reconnect_failed', null);
          }
        },
        onError: (error) {
          _emitLocal('error', error);
        },
      );

      // onDone is triggered after successful connection
      connected = true;
      disconnected = false;
      _connecting = false;
      _reconnectAttempts = 0;
      _reconnectTimer?.cancel();
      _reconnectTimer = null;

      _flushPendingMessages();
      _rejoinRooms();
      _emitLocal('connect', null);
    } catch (err) {
      print('WebSocket connection failed: $err');
      _scheduleReconnect();
    }
  }

  void _scheduleReconnect() {
    if (_reconnectTimer != null) return;

    final delay = (_opts['reconnectionDelay'] * (1 << _reconnectAttempts))
        .clamp(0, _opts['reconnectionDelayMax']);

    _reconnectAttempts++;
    _emitLocal('reconnecting', {'attempt': _reconnectAttempts, 'delay': delay});

    _reconnectTimer = Timer(Duration(milliseconds: delay), () {
      _reconnectTimer = null;
      connect();
    });
  }

  void _rejoinRooms() {
    if (_rooms.isNotEmpty) {
      final rooms = _rooms.toList();
      _send({
        'type': 'rejoinRooms',
        'data': {'rooms': rooms},
      });
    }
  }

  void _flushPendingMessages() {
    while (_pendingMessages.isNotEmpty && connected) {
      final pending = _pendingMessages.removeAt(0);
      _send(pending.message, pending.ack);
    }
  }

  // === Public API ===

  YekongaSocket on(String event, EventCallback callback) {
    _eventListeners.putIfAbsent(event, () => []);
    _eventListeners[event]!.add(callback);
    return this;
  }

  YekongaSocket off(String event, [EventCallback? callback]) {
    if (callback == null) {
      _eventListeners.remove(event);
      return this;
    }

    final listeners = _eventListeners[event];
    if (listeners != null) {
      listeners.remove(callback);
    }
    return this;
  }

  YekongaSocket once(String event, EventCallback callback) {
    void wrapper(data) {
      callback(data);
      off(event, wrapper);
    }

    return on(event, wrapper);
  }

  YekongaSocket emit(String event, [dynamic data, AckCallback? ack]) {
    final msgId = ack != null ? _generateMsgId() : null;

    final message = {
      'type': 'emit',
      if (msgId != null) 'msgId': msgId,
      'data': {'event': event, 'payload': data},
    };

    _send(message, ack);
    return this;
  }

  /// Fluent room emitter: socket.to('room').emit(...)
  RoomEmitter to(String room) {
    return RoomEmitter(this, room);
  }

  /// Direct emit to room: socket.to('room', 'event', data, ack)
  YekongaSocket toRoom(
    String room,
    String event, [
    dynamic data,
    AckCallback? ack,
  ]) {
    final msgId = ack != null ? _generateMsgId() : null;

    final message = {
      'type': 'toRoom',
      if (msgId != null) 'msgId': msgId,
      'data': {'room': room, 'event': event, 'payload': data},
    };

    _send(message, ack);
    return this;
  }

  YekongaSocket join(String room) {
    if (!_rooms.contains(room)) {
      _rooms.add(room);
      _send({
        'type': 'join',
        'data': {'room': room},
      });
    }
    return this;
  }

  YekongaSocket leave(String room) {
    if (_rooms.remove(room)) {
      _send({
        'type': 'leave',
        'data': {'room': room},
      });
    }
    return this;
  }

  YekongaSocket createChannel(String channel) {
    _send({
      'type': 'createChannel',
      'data': {'channel': channel},
    });
    return join(channel);
  }

  void disconnect() {
    _opts['reconnection'] = false;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _channel?.sink.close(status.goingAway);
  }

  // === Private helpers ===

  void _emitLocal(String event, [dynamic data]) {
    final listeners = _eventListeners[event] ?? [];
    for (final cb in listeners) {
      cb(data);
    }
  }

  String _generateMsgId() {
    return 'c_${DateTime.now().millisecondsSinceEpoch}_${_nextMsgId++}';
  }

  void _send(Map<String, dynamic> message, [AckCallback? ack]) {
    final isOpen =
        _channel != null &&
        _channel!.closeCode == null; // sink is open if no close code

    if (!connected || !isOpen) {
      if (ack != null || message['msgId'] != null) {
        _pendingMessages.add(_PendingMessage(message, ack));
      }
      return;
    }

    if (message['msgId'] != null && ack != null) {
      final timer = Timer(Duration(milliseconds: _opts['timeout']), () {
        if (_pendingAcks.containsKey(message['msgId'])) {
          ack(Error('Acknowledgment timeout'));
          _pendingAcks.remove(message['msgId']);
        }
      });

      _pendingAcks[message['msgId']] = _PendingAck(ack, timer);
    }

    try {
      _channel!.sink.add(jsonEncode(message));
    } catch (err) {
      print('Failed to send message: $err');
      if (ack != null || message['msgId'] != null) {
        _pendingMessages.add(_PendingMessage(message, ack));
      }
    }
  }
}

/// Fluent room emitter
class RoomEmitter {
  final YekongaSocket socket;
  final String room;

  RoomEmitter(this.socket, this.room);

  RoomEmitter emit(String event, [dynamic data, AckCallback? ack]) {
    socket.toRoom(room, event, data, ack);
    return this;
  }
}

/// Internal classes
class _PendingMessage {
  final Map<String, dynamic> message;
  final AckCallback? ack;

  _PendingMessage(this.message, this.ack);
}

class _PendingAck {
  final AckCallback ack;
  final Timer timer;

  _PendingAck(this.ack, this.timer);
}

// Example usage:
/*
final socket = YekongaSocket('wss://example.com/chat');

socket
  .on('connect', () => print('Connected as ${socket.id}'))
  .on('disconnect', () => print('Disconnected'))
  .on('message', (msg) => print('Server: $msg'));

socket.emit('chat_message', 'Hello!', (ack) {
  print('Ack: $ack');
});

socket.join('general');
socket.to('general').emit('chat_message', 'Hi room!');

socket.once('welcome', (data) {
  print('Welcome: $data');
});
*/

File: server/static/sdk/yekonga.io.js
/**
 * Browser JavaScript Client (ES5 compatible - works in all modern and older browsers)
 * - Socket.IO-like API: rooms, acknowledgments, reconnection
 * - Exponential backoff reconnection
 * - Message queuing when disconnected
 * - Fluent interface: socket.to('room').emit(...)
 * - No ES6+ features (no let/const, arrow functions, classes with #private, etc.)
 */

function YekongaSocket(url, options) {
    if (!(this instanceof YekongaSocket)) {
        return new YekongaSocket(url, options);
    }

    // Private variables (using naming convention)
    this._ws = null;
    this._url = url;
    this._opts = {
        reconnection: true,
        reconnectionAttempts: Infinity,
        reconnectionDelay: 1000,
        reconnectionDelayMax: 5000,
        timeout: 10000,
    };

    // Merge options
    for (var key in options) {
        if (options.hasOwnProperty(key)) {
            this._opts[key] = options[key];
        }
    }

    this._eventListeners = {};
    this._rooms = {};
    this._pendingMessages = [];
    this._pendingAcks = {};

    this._reconnectAttempts = 0;
    this._reconnectTimer = null;
    this._connecting = false;

    this._nextMsgId = 0;

    // Public properties
    this.id = undefined;
    this.connected = false;
    this.disconnected = true;

    this.connect();
}

YekongaSocket.prototype.connect = function () {
    if (this._connecting) return;
    this._connecting = true;

    try {
        var u = new URL(this._url);
        var socketUrl = u.origin + '/yekonga.io/?ns=' + u.pathname;

        this._ws = new WebSocket(socketUrl);

        this._ws.onopen = function () {
            this.connected = true;
            this.disconnected = false;
            this._connecting = false;
            this._reconnectAttempts = 0;

            if (this._reconnectTimer) {
                clearTimeout(this._reconnectTimer);
                this._reconnectTimer = null;
            }

            this._flushPendingMessages();
            this._rejoinRooms();
            this._emitLocal('connect');
        }.bind(this);

        this._ws.onclose = function (event) {
            this.connected = false;
            this.disconnected = true;
            this._connecting = false;
            this.id = undefined;

            this._emitLocal('disconnect', { code: event.code, reason: event.reason });

            if (this._opts.reconnection && this._reconnectAttempts < this._opts.reconnectionAttempts) {
                this._scheduleReconnect();
            } else if (this._reconnectAttempts >= this._opts.reconnectionAttempts) {
                this._emitLocal('reconnect_failed');
            }
        }.bind(this);

        this._ws.onerror = function (error) {
            this._emitLocal('error', error);
        }.bind(this);

        this._ws.onmessage = function (event) {
            try {
                var msg = JSON.parse(event.data);

                // Handle acknowledgment
                if (msg.type === 'ack' && msg.msgId !== undefined) {
                    var pending = this._pendingAcks[msg.msgId];
                    if (pending) {
                        clearTimeout(pending.timeout);
                        pending.ack(msg.payload !== undefined ? msg.payload : null);
                        delete this._pendingAcks[msg.msgId];
                    }
                    return;
                }

                // Client ID from server
                if (msg.event === 'id') {
                    this.id = msg.data;
                    this._emitLocal('id', this.id);
                    return;
                }

                // Regular event
                if (msg.event) {
                    this._emitLocal(msg.event, msg.data);
                }
            } catch (err) {
                console.error('Failed to parse message:', err);
            }
        }.bind(this);
    } catch (err) {
        console.error('WebSocket connection failed:', err);
        this._scheduleReconnect();
    }
};

YekongaSocket.prototype._scheduleReconnect = function () {
    if (this._reconnectTimer) return;

    var delay = Math.min(
        this._opts.reconnectionDelay * Math.pow(2, this._reconnectAttempts),
        this._opts.reconnectionDelayMax
    );

    this._reconnectAttempts++;
    this._emitLocal('reconnecting', { attempt: this._reconnectAttempts, delay: delay });

    this._reconnectTimer = setTimeout(function () {
        this._reconnectTimer = null;
        this.connect();
    }.bind(this), delay);
};

YekongaSocket.prototype._rejoinRooms = function () {
    var rooms = [];
    for (var room in this._rooms) {
        if (this._rooms.hasOwnProperty(room)) {
            rooms.push(room);
        }
    }
    if (rooms.length > 0) {
        this._send({ type: 'rejoinRooms', data: { rooms: rooms } });
    }
};

YekongaSocket.prototype._flushPendingMessages = function () {
    while (this._pendingMessages.length > 0 && this.connected) {
        var item = this._pendingMessages.shift();
        this._send(item.message, item.ack);
    }
};

// === Public API ===

YekongaSocket.prototype.on = function (event, callback) {
    if (!this._eventListeners[event]) {
        this._eventListeners[event] = [];
    }
    this._eventListeners[event].push(callback);
    return this;
};

YekongaSocket.prototype.off = function (event, callback) {
    if (!callback) {
        delete this._eventListeners[event];
        return this;
    }

    var listeners = this._eventListeners[event];
    if (listeners) {
        var newListeners = [];
        for (var i = 0; i < listeners.length; i++) {
            if (listeners[i] !== callback) {
                newListeners.push(listeners[i]);
            }
        }
        this._eventListeners[event] = newListeners;
    }
    return this;
};

YekongaSocket.prototype.once = function (event, callback) {
    var self = this;
    function wrapper(data) {
        callback(data);
        self.off(event, wrapper);
    }
    return this.on(event, wrapper);
};

YekongaSocket.prototype.emit = function (event, data, ack) {
    // Handle: emit(event), emit(event, data), emit(event, data, ack)
    if (typeof data === 'function') {
        ack = data;
        data = undefined;
    }

    var msgId = ack ? this._generateMsgId() : undefined;

    var message = {
        type: 'emit',
        msgId: msgId,
        data: { event: event, payload: data }
    };

    this._send(message, ack);
    return this;
};

YekongaSocket.prototype.to = function (room) {
    var args = Array.prototype.slice.call(arguments);

    // If only room is provided → return fluent interface
    if (args.length === 1) {
        var self = this;
        return {
            emit: function (event, data, ack) {
                return self.to(room, event, data, ack);
            }
        };
    }

    // to(room, event, [data], [ack])
    var event = args[1];
    var data = args[2];
    var ack = args[3];

    if (typeof data === 'function') {
        ack = data;
        data = undefined;
    }

    var msgId = ack ? this._generateMsgId() : undefined;

    var message = {
        type: 'toRoom',
        msgId: msgId,
        data: { room: room, event: event, payload: data }
    };

    this._send(message, ack);
    return this;
};

YekongaSocket.prototype.join = function (room) {
    if (!this._rooms[room]) {
        this._rooms[room] = true;
        this._send({ type: 'join', data: { room: room } });
    }
    return this;
};

YekongaSocket.prototype.leave = function (room) {
    if (this._rooms[room]) {
        delete this._rooms[room];
        this._send({ type: 'leave', data: { room: room } });
    }
    return this;
};

YekongaSocket.prototype.createChannel = function (channel) {
    this._send({ type: 'createChannel', data: { channel: channel } });
    return this.join(channel);
};

YekongaSocket.prototype.disconnect = function () {
    this._opts.reconnection = false;
    if (this._reconnectTimer) {
        clearTimeout(this._reconnectTimer);
        this._reconnectTimer = null;
    }
    if (this._ws) {
        this._ws.close();
    }
    return this;
};

// === Private helpers ===

YekongaSocket.prototype._emitLocal = function (event, data) {
    var listeners = this._eventListeners[event] || [];
    for (var i = 0; i < listeners.length; i++) {
        listeners[i](data);
    }
};

YekongaSocket.prototype._generateMsgId = function () {
    return 'c_' + Date.now() + '_' + this._nextMsgId++;
};

YekongaSocket.prototype._send = function (message, ack) {
    var isOpen = this._ws && this._ws.readyState === WebSocket.OPEN;

    if (!this.connected || !isOpen) {
        if (ack || message.msgId) {
            this._pendingMessages.push({ message: message, ack: ack });
        }
        return;
    }

    if (message.msgId && ack) {
        var timeout = setTimeout(function () {
            if (this._pendingAcks[message.msgId]) {
                ack(new Error('Acknowledgment timeout'));
                delete this._pendingAcks[message.msgId];
            }
        }.bind(this), this._opts.timeout);

        this._pendingAcks[message.msgId] = { ack: ack, timeout: timeout };
    }

    try {
        this._ws.send(JSON.stringify(message));
    } catch (err) {
        console.warn('Failed to send message:', err);
        if (ack || message.msgId) {
            this._pendingMessages.push({ message: message, ack: ack });
        }
    }
};

// Example usage:
/*
var socket = new YekongaSocket('wss://example.com/chat');

socket
  .on('connect', function() { console.log('Connected as', socket.id); })
  .on('disconnect', function() { console.log('Disconnected'); })
  .on('message', function(msg) { console.log('Server says:', msg); });

socket.emit('chat_message', 'Hello!', function(ack) {
  console.log('Server acknowledged:', ack);
});

socket.join('general');
socket.to('general').emit('chat_message', 'Hi everyone!');

socket.once('welcome', function(data) {
  console.log('Welcome:', data);
});
*/

if (typeof module !== 'undefined' && module.exports) {
    module.exports = YekongaSocket;
} else if (typeof window !== 'undefined') {
    window.YekongaSocket = YekongaSocket;
}